core/src/main/java/org/elasticsearch/action/index/TransportIndexAction.java

```diff
@@ -175,29 +182,40 @@ protected Location onReplicaShard(IndexRequest request, IndexShard indexShard) {
     }
 
     public static WriteResult<IndexResponse> executeIndexRequestOnPrimary(IndexRequest request, IndexShard indexShard,
-            MappingUpdatedAction mappingUpdatedAction) throws Exception {
-        Engine.Index operation = prepareIndexOperationOnPrimary(request, indexShard);
+            MappingUpdatedAction mappingUpdatedAction) {
+        Engine.Index operation;
+        try {
+            operation = prepareIndexOperationOnPrimary(request, indexShard);
+        } catch (Exception e) {
+            return new WriteResult<>(e);
+        }
         Mapping update = operation.parsedDoc().dynamicMappingsUpdate();
         final ShardId shardId = indexShard.shardId();
         if (update != null) {
-            mappingUpdatedAction.updateMappingOnMaster(shardId.getIndex(), request.type(), update);
-            operation = prepareIndexOperationOnPrimary(request, indexShard);
+            try {
+                mappingUpdatedAction.updateMappingOnMaster(shardId.getIndex(), request.type(), update);
+                operation = prepareIndexOperationOnPrimary(request, indexShard);
+            } catch (Exception e) {
```

Comment:
I'm on the fence regarding these - originally I was thinking we should limit the failure handling to things that happened on the engine level. Not saying this is bad but I want to hear why you decided to add this.